ComplexSqlPredicateTest

Holmistr opened the pull request.

devOpsHazelcast commented:
Test FAILed.

pveentjer commented on tests/tests-hz/src/main/java/com/hazelcast/simulator/tests/map/ComplexSqlPredicateTest.java:
```diff
@@ -0,0 +1,545 @@
+    @Setup
+    public void setup() {
+        this.map = targetInstance.getMap(name);
+        generatedIds = generateStrings(keyCount, idLength);
```
This could lead to a huge amount of data being kept in memory; even though it is just readonly.

Holmistr replied:
I understand, but I'm thinking whether it is a problem or how it should be solved. I need to store the IDs somewhere to be able to choose from them.

pveentjer replied:
It is a problem which should be solved because it can impact the measurements.

You don't need to store the id's, you can generate them on the fly.

Holmistr replied:
Please, just ignore my previous comment, I probably suffered from temporal brain damage. I figured it out.

pveentjer commented on tests/tests-hz/src/main/java/com/hazelcast/simulator/tests/map/ComplexSqlPredicateTest.java:
```diff
@@ -0,0 +1,545 @@
+    @TimeStep(prob = 0.4)
+    public void searchSingle(ThreadState state) {
+        String predicate = "id = '" + generatedIds[state.random.nextInt(generatedIds.length)] + "'";
+        System.out.println( map.values(new SqlPredicate(predicate)));
```
I hope you are going to remove system.out from the code :)

Holmistr replied:
Oh, leftover from testing :)

pveentjer commented on tests/tests-hz/src/main/java/com/hazelcast/simulator/tests/map/ComplexSqlPredicateTest.java:
```diff
@@ -0,0 +1,545 @@
+    @TimeStep(prob = 0.1)
+    public void searchHalf(ThreadState state) {
+        String predicate = "actionCode = " + (state.random.nextBoolean() ? ActionCode.CODE_0000.toCode() : ActionCode.CODE_9280.toCode());
+        map.values(new SqlPredicate(predicate));
```
You are allowed to return the value in a timestep method. Currently it is ignored; but in the future I'll add dead code elimination protection.

Holmistr replied:
Ok

pveentjer commented on tests/tests-hz/src/main/java/com/hazelcast/simulator/tests/map/ComplexSqlPredicateTest.java:
```diff
@@ -0,0 +1,545 @@
+        map.values(new SqlPredicate(predicate));
+    }
+
+    public class ThreadState extends BaseThreadState {
```
Random is already available on BaseThreadState.

Holmistr replied:
Nice.

pveentjer commented on tests/tests-hz/src/main/java/com/hazelcast/simulator/tests/map/ComplexSqlPredicateTest.java:
```diff
@@ -0,0 +1,545 @@
+        logger.info(getOperationCountInformation(targetInstance));
+    }
+
+    public static class Transaction implements Portable, Serializable {
```
I would remove Seriazable.

Holmistr replied:
Ok

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test FAILed.

Holmistr commented:
@pveentjer Thanks for the review, I have addressed all of your comments. I had to change the checkstyle.xml in order to pass the check, let me know if you think it's acceptable. My justifications are:
* the predicate lines would become much less readable if I split them (I tried) -> increased line size
* parameter count for constructor -> I would have to use setters -> a lot more code
* method count -> I know I'm not using all the getters/setters right now, but I can imagine the situation where it could be useful

devOpsHazelcast commented:
Test PASSed.

Holmistr commented:
@pveentjer I have also added update method.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test FAILed.

pveentjer commented on checkstyle/checkstyle.xml:
```diff
@@ -140,7 +140,7 @@
         <!-- See http://checkstyle.sf.net/config_sizes.html -->
         <module name="LineLength">
-            <property name="max" value="130"/>
+            <property name="max" value="140"/>
```
It is best to ignore the test for checkstyle.

devOpsHazelcast commented:
Test PASSed.

Holmistr commented:
@pveentjer I addressed all your comments.

devOpsHazelcast commented:
Test PASSed.

Holmistr commented:
The test actually still has some problems. When using multiple timestep method (not just singleSearch), I get NullPointerExceptions. I honestly have no idea why. When I investigated it, I found out that right after the setup loop with streamer.pushEntry, if I do another for loop with gets, I see that some of the keys are missing. I believe I'm missing something fundamental in Simulator's design, @pveentjer , any idea?

pveentjer commented:
Closing this since it is 3 years old.

Feel free to reopen the PR if it still needs to be merged
